OCS deployment with ACM  and DR workload

shylesh opened the pull request:
1.odf multiclustr orchestrator deployment and config
2.mirror peer configuration and deployment for rbd
3.openshift dr-cluster and dr-hub operator deployment and config
4.volumereplication class creation
5.odr secret creation
6. DR workload

Signed-off-by: Shylesh Kumar Mohan <shmohan@redhat.com>

prsurve commented on conf/ocsci/submariner_upstream.yaml:
```diff
@@ -0,0 +1,9 @@
+  submariner_source: "upstream"
+  submariner_url: "https://get.submariner.io"
+  submariner_info_file: "broker-info.subm"
+  submariner_default_gateway_node: "compute-1"
```
in case of ipi we will not have `compute-1` we need to set this value based on node name

prsurve replied:
yes we need only one node. Can be any any node from worker node

shylesh replied:
done, now fetching the first node in the workers list

prsurve commented on ocs_ci/deployment/acm.py:
```diff
@@ -0,0 +1,198 @@
+            # Copy submariner from ~/.local/bin to ocs-ci/bin
+            # ~/.local/bin is the default path selected by submariner script
+            # TODO: create symlink
+            cp_cmd = f"cp ~/.local/bin/subctl {config.RUN['bin_dir']}"
```
can we do a mv here ?

shylesh replied:
done

prsurve commented on ocs_ci/templates/ocs-deployment/multicluster/openshift_dr_cluster_operator.yaml:
```diff
@@ -0,0 +1,33 @@
+apiVersion: v1
+data:
+  ramen_manager_config.yaml: |
+    apiVersion: ramendr.openshift.io/v1alpha1
+    kind: RamenConfig
+    leaderElection:
+      resourceName: dr-cluster.ramendr.openshift.io
+    ramenControllerType: "dr-cluster"
+    s3StoreProfiles:
+      - s3ProfileName: odr-s3secret-east
+        s3CompatibleEndpoint: https://s3.us-east-2.amazonaws.com:443
+        s3Region: us-east-2
+        s3SecretRef:
+          name: odr-s3secret-east
+          namespace: openshift-dr-system
+      - s3ProfileName: odr-s3secret-west
+        s3CompatibleEndpoint: https://s3.us-east-2.amazonaws.com:443
+        s3Region: us-east-2
+        s3SecretRef:
+          name: odr-s3secret-west
+          namespace: openshift-dr-system
```
There is a plan to use MCG only and not use AWS

shylesh replied:
@prsurve I can handle this in next PR. Also, I need to get handson for how to configure MCG for this purpose. is that fine ?

prsurve commented on ocs_ci/templates/ocs-deployment/multicluster/openshift_dr_hub_operator.yaml:
```diff
@@ -0,0 +1,34 @@
+apiVersion: v1
+data:
+  ramen_manager_config.yaml: |
+    apiVersion: ramendr.openshift.io/v1alpha1
+    kind: RamenConfig
+    leaderElection:
+      leaderElect: true
+      resourceName: hub.ramendr.openshift.io
+    ramenControllerType: "dr-hub"
+    s3StoreProfiles:
+      - s3ProfileName: odr-s3secret-east
+        s3CompatibleEndpoint: https://s3.us-east-2.amazonaws.com:443
+        s3Region: us-east-2
+        s3SecretRef:
+          name: odr-s3secret-east
+          namespace: openshift-dr-system
+      - s3ProfileName: odr-s3secret-west
+        s3CompatibleEndpoint: https://s3.us-east-2.amazonaws.com:443
+        s3Region: us-east-2
+        s3SecretRef:
+          name: odr-s3secret-west
+          namespace: openshift-dr-system
```
Same

prsurve commented on ocs_ci/deployment/deployment.py:
```diff
@@ -169,22 +178,149 @@ def deploy_cluster(self, log_cli_level="DEBUG"):
+            kind='MirrorPeer', 
+            namespace=constants.dr_default_namespace
+        )
+        mirror_peer.wait_for_phase(phase="ExchangingSecret")
```
we should have 2 condition some time we also see `ExchangedSecret`

shylesh replied:
I have added both condition one after the other, will it be serial in changing the states OR we can expect either one of this ?

prsurve commented on conf/ocsci/dr_workload.yaml:
```diff
@@ -0,0 +1,2 @@
+ENV_DATA:
+  dr_workload_repo_url: "https://github.com/shylesh/ocm-ramen-samples.git"
```
this repo will only work on the given  cluster name if diff app are not deployed https://github.com/shylesh/ocm-ramen-samples/blob/main/subscriptions/busybox/drpc.yaml#L9

shylesh replied:
@prsurve Waiting for clarity on this from Shyam.

shylesh replied:
We will handle this as and when required.

petr-balogh commented on ocs_ci/utility/utils.py:
```diff
@@ -458,6 +458,60 @@ def run_cmd(cmd, secrets=None, timeout=600, ignore_error=False, **kwargs):
+    """
+    # Skip indexed cluster while running commands
+    # Useful to skip operations on ACM cluster
+    skip_index = kwargs.get("skip_index", None)
```
I think that skip_index should be regular parameter of the function. I think it should not be then passed to exec_cmd as if you have it in kwargs and do not pop that parameter from kwargs it will be passed to exec_cmd.

shylesh replied:
done

petr-balogh commented on ocs_ci/deployment/deployment.py:
```diff
@@ -174,23 +196,296 @@ def deploy_cluster(self, log_cli_level="DEBUG"):
             logger.warning("OCS deployment will be skipped")
 
+        # Multicluster: Install ODF multicluster orchestrator only on hub
+        if config.multicluster:
```
I think that this code for multicluster should be moved out of the deploy_cluster method and you should just call the new method you will move the code to from here.

shylesh replied:
done, separated the class for multicluster dr related deployment code.

petr-balogh commented on ocs_ci/deployment/deployment.py:
```diff
@@ -174,23 +196,296 @@ def deploy_cluster(self, log_cli_level="DEBUG"):
-                    collect_ocs_logs("deployment", ocp=False)
-                raise
+            for i in range(config.clusters):
+                if config.multicluster:
+                    if config.acm_index == i:
+                        continue
```
```suggestion
                if config.multicluster and config.acm_index == i:
                        continue
```

shylesh replied:
done

petr-balogh commented on ocs_ci/deployment/deployment.py:
```diff
@@ -174,23 +196,296 @@ def deploy_cluster(self, log_cli_level="DEBUG"):
+
+        """
+        # For first cut just returning east and west
+        return ["east", "west"]
```
This we need to take out of configuration.

shylesh replied:
I will update it in the next PR as MCG will be the default metadata object store.

petr-balogh commented on ocs_ci/ocs/constants.py:
```diff
@@ -619,6 +620,42 @@
+ODR_S3_SECRET_YAML = os.path.join(TEMPLATE_MULTICLUSTER_DIR, "odr_s3_secret.yaml")
+
+# DR constants
+dr_default_namespace = "openshift-dr-systems"
```
For constants we are following upper case for constants names but all your defined constants are lower case. Please follow the standard here.

shylesh replied:
done, all constants are now uppercase

petr-balogh commented on ocs_ci/deployment/acm.py:
```diff
@@ -0,0 +1,192 @@
+
+        # Join all the clusters (except ACM cluster in case of hub deployment)
+        for cluster in config.clusters:
+            cluster_index = config.clusters.index(cluster)
```
I see you are using this config.cluster.index(cluster) construction on a few places in the code.
What about to add the index to the cluster configuration directly when we are creating multi cluster configuration. I think that we don't change order so once we init it we should know the index and we could probably set it some of the section. Maybe you can have some method in configuration object for cluster which will hold its index value. WDYT?

shylesh replied:
Done, now we are storing a map of Config object and its index as part of MulticlusterConfig object.

petr-balogh commented on ocs_ci/deployment/acm.py:
```diff
@@ -0,0 +1,192 @@
+
+    """
+    cmd = " ".join("subctl", cmd)
+    out = run_cmd(cmd)
+    logger.info(out)
```
return run_cmd(cmd)

logging is done in run_cmd I think

shylesh replied:
done

petr-balogh commented on ocs_ci/deployment/acm.py:
```diff
@@ -0,0 +1,192 @@
+        kubeconf_list = []
+        for i in self.dr_only_list:
+            kubeconf_list.append(config.clusters[i].RUN["kubeconfig"])
+        connct_check = "verify " + " ".join(kubeconf_list) + "--only connectivity"
```
string formating

shylesh replied:
done

shylesh commented:
Rebased on latest master

shylesh commented:
Rebased on latest master

shylesh commented on ocs_ci/deployment/acm.py:
```diff
@@ -0,0 +1,192 @@
+    Args:
+        cmd: subctl command to be executed
+
+    Returns:
```
done

shylesh commented on ocs_ci/deployment/acm.py:
```diff
@@ -0,0 +1,192 @@
+    logger.info(out)
+
+
+def run_subctl_cmd_interactive(cmd, prompt, answer):
```
done

shylesh commented on ocs_ci/deployment/acm.py:
```diff
@@ -0,0 +1,192 @@
+        self.submariner_configure_upstream()
+
+    def deploy_downstream(self):
+        pass
```
done

shylesh commented on ocs_ci/deployment/acm.py:
```diff
@@ -0,0 +1,192 @@
+            try:
+                run_cmd(download_cmd)
+            except CommandFailed:
+                logger.exception("Failed to download submariner binary")
```
done, handled both download and run failures.

shylesh replied:
@Avilir , there are 2 parts to this function
1. Curl failures: may be because of network down or something
2. curl will actually fetch a downloader script which will then download submariner binary

So I hope both exceptions convey it. 
Note: We are not running submariner command here but just the download

shylesh commented on ocs_ci/deployment/acm.py:
```diff
@@ -0,0 +1,192 @@
+        # Join all the clusters (except ACM cluster in case of hub deployment)
+        for cluster in config.clusters:
+            cluster_index = config.clusters.index(cluster)
+            if not cluster_index == config.acm_index:
```
done

shylesh commented on ocs_ci/deployment/deployment.py:
```diff
@@ -174,23 +196,296 @@ def deploy_cluster(self, log_cli_level="DEBUG"):
+            # Need to run on all the clusters
+            for i in range(config.nclusters):
+                config.switch_ctx(i)
+                disable_default_sources()
```
removed it, Pratik pointed to me that its been handled as part OCS deployment

shylesh commented on ocs_ci/deployment/deployment.py:
```diff
@@ -174,23 +196,296 @@ def deploy_cluster(self, log_cli_level="DEBUG"):
+                        continue
+                    else:
+                        config.switch_ctx(i)
+                        ocs_registry_image = config.DEPLOYMENT["ocs_registry_image"]
```
done

shylesh commented on ocs_ci/deployment/deployment.py:
```diff
@@ -174,23 +195,67 @@ def deploy_cluster(self, log_cli_level="DEBUG"):
+            deploy_dr = MultiClusterDROperatorsDeploy(dr_conf)
+            deploy_dr.deploy()
+
+    def get_dr_conf(self):
```
Rename the function to `get_rdr_conf`

shylesh commented on ocs_ci/deployment/deployment.py:
```diff
@@ -1198,3 +1263,303 @@ def setup_persistent_monitoring():
+        self.cephfs_dr = dr_conf.get("cephfs_dr_scenario", False)
+        self.meta_map = {"s3": self.s3_meta_obj_store, "mcg": self.mcg_meta_obj_store}
+        # Default to s3 for metadata store
+        self.meta_obj_store = dr_conf.get("dr_metadata_store", "s3")
```
be specific to `awss3`

shylesh replied:
done

shylesh commented on ocs_ci/deployment/deployment.py:
```diff
@@ -1198,3 +1263,303 @@ def setup_persistent_monitoring():
+            mode="w+", prefix="dr_policy_hub_", delete=False
+        )
+        templating.dump_to_temp_yaml(dr_policy_hub_data, dr_policy_hub_yaml.name)
+        run_cmd(f"oc create -f {dr_policy_hub_yaml.name}")
```
Add status validation to check for object store availability

shylesh replied:
I will handle this in next PR

shylesh commented on ocs_ci/ocs/dr/dr_workload.py:
```diff
@@ -0,0 +1,79 @@
+            os.path.join(self.target_clone_dir, constants.DR_WORKLOAD_REPO_BASE_DIR),
+            "subscription",
+        )
+        run_cmd(f"oc create -k {workload_subscription_dir}")
```
Add teardown for workloads as well `oc delete`

shylesh replied:
I will take up this in next pr

shylesh commented on ocs_ci/ocs/utils.py:
```diff
@@ -1203,3 +1203,34 @@ def enable_console_plugin():
+
+    """
+    non_acm_list = []
+    for i in len(ocsci_config.clusters):
```
my bad, changed it. Thanks.

shylesh replied:
done

shylesh commented on ocs_ci/utility/utils.py:
```diff
@@ -458,6 +460,85 @@ def run_cmd(cmd, secrets=None, timeout=600, ignore_error=False, **kwargs):
+
+    """
+    child = pexpect.spawn(cmd)
+    for prompt, answer in zip(prompts, answers):
```
nice idea, done.

shylesh replied:
done

shylesh commented:
Rebased on latest master

jimdevops19 commented on ocs_ci/deployment/deployment.py:
```diff
@@ -174,23 +195,67 @@ def deploy_cluster(self, log_cli_level="DEBUG"):
+                    raise
+            config.reset_ctx()
+
+            # Run ocs_install_verification here only in case of multiclsuter.
```
```suggestion
            # Run ocs_install_verification here only in case of multicluster.
```

shylesh replied:
done

jimdevops19 commented on ocs_ci/templates/ocs-deployment/multicluster/dr_policy_acm_hub.yaml:
```diff
@@ -0,0 +1,12 @@
+  - name: PLACE_HOLDER_CLUSTER1 
+    s3ProfileName: odr-s3-secret-east 
+  - name: PLACE_HOLDER_CLUSTER2 
+    s3ProfileName: odr-s3-secert-west 
```
```suggestion
    s3ProfileName: odr-s3-secret-west 
```

shylesh replied:
done

shylesh commented on ocs_ci/deployment/deployment.py:
```diff
@@ -1198,3 +1263,308 @@ def setup_persistent_monitoring():
+
+    def __init__(self, dr_conf):
+        # DR use case could be RBD or CephFS or Both
+        self.rbd_dr = dr_conf.get("rbd_dr_scenario", False)
```
Change `rbd_dr` to just `rbd` as context is already DR

shylesh replied:
done

shylesh commented:
Rebased on latest master

sidhant-agrawal commented on ocs_ci/deployment/deployment.py:
```diff
@@ -1198,3 +1263,308 @@ def setup_persistent_monitoring():
+        )
+        # Need to satisfy both the phases
+        mirror_peer.wait_for_phase(phase="ExchangingSecret")
+        mirror_peer.wait_for_phase(phase="ExchangedSecret")
```
IMO we should not wait for both in serial order here.
Either one of `ExchangedSecret` and `ExchangingSecret` is an acceptable result.

shylesh replied:
done, now we are waiting for either one of them

sidhant-agrawal commented on ocs_ci/templates/ocs-deployment/multicluster/odf_multicluster_orchestrator.yaml:
```diff
@@ -0,0 +1,17 @@
+  name: odf-multicluster-orchestrator
+  source:  ocs-catalogsource
+  sourceNamespace: openshift-marketplace
+  currentCSV: odf-multicluster-orchestrator.v4.9.0-120.ci
```
Specific version mentioned here is old now. Needs to be updated.

shylesh replied:
I can add code for version modification in the next PR.

shylesh replied:
marking as TODO

sidhant-agrawal commented on ocs_ci/templates/ocs-deployment/multicluster/odr_s3_secret.yaml:
```diff
@@ -0,0 +1,10 @@
+metadata:
+  # Replace PLACE_HOLDER_FOR_REGION with region name(east, west) etc
+  name: odr-s3secret-PLACE_HOLDER_FOR_REGION
+  namespace: odr-system
```
`odr-system` to be replaced by `openshift-dr-system`

shylesh replied:
done

shylesh commented on ocs_ci/ocs/constants.py:
```diff
@@ -596,6 +597,43 @@
+# DR constants
+SUBMARINER_DOWNLOAD_URL = "https://get.submariner.io"
+DR_DEFAULT_NAMESPACE = "openshift-dr-systems"
+TOKEN_EXCHANGE_AGENT_LABEL = "token-exchange-agent "
```
sorry, that was a mistake.done

prsurve commented on ocs_ci/deployment/acm.py:
```diff
@@ -0,0 +1,204 @@
+                    f"--clusterid c{self.cluster_seq} --natt=false"
+                )
+                try:
+                    # run_subctl_cmd(
+                    #     join_cmd,
+                    # )
```
looks like code is commented

shylesh replied:
@prsurve , yep, sorry, I was testing something incrementally so I had it commented and I wanted another branch for Jim's testing so pushed this as is. Soon I will uncomment this.

shylesh replied:
Uncommented the proper lines

shylesh commented:
Rebased on latest master and resolved conflicts.

shylesh commented:
> PR validation
> 
> Cluster Name: Cluster Configuration: PR Test Suite: acceptance PR Test Path: tests/ Additional Test Params: OCP VERSION: 4.9-ga OCS VERSION: 4.9 tested against branch: master
> 
> [Job UNSTABLE](https://ocs4-jenkins-csb-odf-qe.apps.ocp-c1.prod.psi.redhat.com/job/qe-trigger-test-pr/2273//testReport) (some or all tests failed).

All tests passed except 1 which is due to connectivity issues ```Could not connect to the endpoint URL: ```

shylesh commented:
Verified deployment locally for multicluster deployment of ODF and DR operators.

clacroix12 reviewed (dismissed):
Looks good as far as I can tell. I had a few comments but nothing critical so approving.

shylesh commented on ocs_ci/deployment/acm.py:
```diff
@@ -0,0 +1,203 @@
+    Args:
+        cmd(str): Command to be executed
+        prompt(str): Expected question during command run which needs to be provided
+        answer(str): Answer for the prompt
```
done

shylesh commented on ocs_ci/framework/pytest_customization/ocscilib.py:
```diff
@@ -469,9 +469,15 @@ def process_cluster_cli_params(config):
     )
-    ocsci_config.RUN["kubeconfig"] = os.path.join(
-        cluster_path, ocsci_config.RUN["kubeconfig_location"]
+    print(f"Setting KUBECONFIG on {ocsci_config.ENV_DATA['cluster_name']}")
```
sorry, leftover from debugging. removed it

shylesh commented on ocs_ci/ocs/utils.py:
```diff
@@ -1227,3 +1227,34 @@ def enable_console_plugin():
+        list: of cluster config objects
+
+    """
+    non_acm_list = []
+    for i in range(len(ocsci_config.clusters)):
+        if i == ocsci_config.get_acm_index():
+            continue
+        else:
+            non_acm_list.append(ocsci_config.clusters[i])
+
+    return non_acm_list
```
Agree,  This code is leftover from earlier approach. After discussing with you on handling the things within config things have changed and I missed this one. I will take up in the next immediate PR which addresses this issue.

shylesh commented:
@clacroix12 For now I addressed your comments which doesn't require retesting. Only one comment is left unaddressed which I will take up in the next PR.
